File: src/algebra2_radical_expressions_medium.py
"""
algebra2 - radical_expressions (medium)
Generated: 2026-02-22T04:25:57.017724
"""

from problem_utils import *

def generate():
    problem_type = randint(1, 6)

    if problem_type == 1:
        # Simplify sqrt(a^2 * b) where b is prime or small
        base = randint(2, 10)
        inner_factor = choice([2, 3, 5, 7, 11, 13, 17])
        radicand = base**2 * inner_factor
        ans = base * sqrt(inner_factor)

        return problem(
            question=f"Simplify $\\sqrt{{{radicand}}}$",
            answer=ans,
            difficulty=(1300, 1400),
            topic="algebra2/radical_expressions",
            solution=steps(
                f"Factor: $\\sqrt{{{radicand}}} = \\sqrt{{{base}^2 \\cdot {inner_factor}}}$",
                f"$= \\sqrt{{{base}^2}} \\cdot \\sqrt{{{inner_factor}}} = {latex(ans)}$"
            )
        )

    elif problem_type == 2:
        # Add/subtract radicals (may require simplifying first)
        n = choice([2, 3, 5, 6, 7, 10, 11])
        a_coeff = randint(1, 8)
        b_coeff = randint(1, 8)
        sign = choice([1, -1])
        result_coeff = a_coeff + sign * b_coeff
        ans = result_coeff * sqrt(n)

        sign_str = "+" if sign == 1 else "-"
        return problem(
            question=f"Simplify ${a_coeff}\\sqrt{{{n}}} {sign_str} {b_coeff}\\sqrt{{{n}}}$",
            answer=ans,
            difficulty=(1300, 1400),
            topic="algebra2/radical_expressions",
            solution=steps(
                f"Combine like terms (both have $\\sqrt{{{n}}}$)",
                f"$= ({a_coeff} {sign_str} {b_coeff})\\sqrt{{{n}}} = {result_coeff}\\sqrt{{{n}}}$"
            )
        )

    elif problem_type == 3:
        # Multiply radicals: a√m * b√n
        a_val = randint(1, 5)
        b_val = randint(1, 5)
        m_val = randint(2, 12)
        n_val = randint(2, 12)
        product = m_val * n_val
        ans = simplify(a_val * b_val * sqrt(product))

        return problem(
            question=f"Simplify ${a_val}\\sqrt{{{m_val}}} \\cdot {b_val}\\sqrt{{{n_val}}}$",
            answer=ans,
            difficulty=(1400, 1500),
            topic="algebra2/radical_expressions",
            solution=steps(
                f"Multiply coefficients and radicands separately:",
                f"$= ({a_val} \\cdot {b_val})\\sqrt{{{m_val} \\cdot {n_val}}} = {a_val*b_val}\\sqrt{{{product}}}$",
                f"$= {latex(ans)}$"
            )
        )

    elif problem_type == 4:
        # Rationalize denominator: a / sqrt(b)
        a_val = randint(1, 12)
        b_val = choice([2, 3, 5, 6, 7, 10, 11])
        ans = simplify(a_val * sqrt(b_val) / b_val)

        return problem(
            question=f"Rationalize the denominator: $\\frac{{{a_val}}}{{\\sqrt{{{b_val}}}}}$",
            answer=ans,
            difficulty=(1400, 1500),
            topic="algebra2/radical_expressions",
            solution=steps(
                f"Multiply numerator and denominator by $\\sqrt{{{b_val}}}$",
                f"$= \\frac{{{a_val}\\sqrt{{{b_val}}}}}{{{b_val}}}$",
                f"$= {latex(ans)}$"
            )
        )

    elif problem_type == 5:
        # Simplify (√a + √b)^2 or (√a - √b)^2
        a_val = choice([2, 3, 5, 7])
        b_val = choice([2, 3, 5, 7, 11])
        while b_val == a_val:
            b_val = choice([2, 3, 5, 7, 11])
        sign = choice([1, -1])
        sign_str = "+" if sign == 1 else "-"

        ans = simplify((sqrt(a_val) + sign*sqrt(b_val))**2)

        return problem(
            question=f"Expand and simplify $(\\sqrt{{{a_val}}} {sign_str} \\sqrt{{{b_val}}})^2$",
            answer=ans,
            difficulty=(1500, 1600),
            topic="algebra2/radical_expressions",
            solution=steps(
                f"Use $(a {sign_str} b)^2 = a^2 {sign_str} 2ab + b^2$",
                f"$(\\sqrt{{{a_val}}})^2 {sign_str} 2\\sqrt{{{a_val}}}\\sqrt{{{b_val}}} + (\\sqrt{{{b_val}}})^2$",
                f"$= {a_val} {sign_str} 2\\sqrt{{{a_val * b_val}}} + {b_val}$",
                f"$= {latex(ans)}$"
            )
        )

    else:
        # Simplify cube root expression: a * cbrt(b^3 * c)
        base = randint(2, 5)
        inner = choice([2, 3, 5, 7])
        coeff = randint(1, 4)
        radicand = base**3 * inner
        from sympy import cbrt
        ans = simplify(coeff * base * cbrt(inner))

        return problem(
            question=f"Simplify: ${coeff}\\sqrt[3]{{{radicand}}}$",
            answer=ans,
            difficulty=(1450, 1600),
            topic="algebra2/radical_expressions",
            solution=steps(
                f"Factor: $\\sqrt[3]{{{radicand}}} = \\sqrt[3]{{{base}^3 \\cdot {inner}}} = {base}\\sqrt[3]{{{inner}}}$",
                f"${coeff} \\cdot {base}\\sqrt[3]{{{inner}}} = {latex(ans)}$"
            )
        )

emit(generate())

File: src/algebra2_radical_expressions_medium.jl
# algebra2 - radical_expressions (medium)
# Generated: 2026-03-08T20:17:15.926034

include(joinpath(@__DIR__, "..", "julia", "src", "ProblemUtils.jl"))
using .ProblemUtils

@script x begin
    set_topic!("algebra2/radical_expressions")
    
    problem_type = rand(1:6)
    
    if problem_type == 1
        # Simplifying radicals with variables (1200-1400)
        # √(a²b²c) = ab√c
        a = randint(2, 8)
        b = randint(2, 6)
        c = choice([2, 3, 5, 6, 7, 10, 11, 13, 14, 15])
        
        radicand = a^2 * b^2 * c
        ans_coeff = a * b
        
        problem(
            question="Simplify \\(\\sqrt{$(radicand)x^2}\\)",
            answer="$(ans_coeff)x\\sqrt{$(c)}",
            difficulty=(1200, 1350),
            solution=steps(
                sol("Given", "\\sqrt{$(radicand)x^2}"),
                "Factor perfect squares: \\(\\sqrt{$(a^2) \\cdot $(b^2) \\cdot $(c) \\cdot x^2}\\)",
                "Simplify: \\($(a) \\cdot $(b) \\cdot x \\cdot \\sqrt{$(c)} = $(ans_coeff)x\\sqrt{$(c)}\\)",
            ),
            time=90,
        )
        
    elseif problem_type == 2
        # Rationalizing denominators (1300-1500)
        # a/(b√c) = a√c/(bc)
        a = randint(2, 15)
        b = randint(2, 8)
        c = choice([2, 3, 5, 6, 7, 10, 11, 13])
        
        denom = b * c
        gcd_val = gcd(a, denom)
        num_simplified = div(a, gcd_val)
        denom_simplified = div(denom, gcd_val)
        
        if denom_simplified == 1
            ans_str = "$(num_simplified)\\sqrt{$(c)}"
        else
            ans_str = "\\frac{$(num_simplified)\\sqrt{$(c)}}{$(denom_simplified)}"
        end
        
        problem(
            question="Rationalize the denominator: \\(\\frac{$(a)}{$(b)\\sqrt{$(c)}}\\)",
            answer=ans_str,
            difficulty=(1300, 1500),
            solution=steps(
                sol("Given", "\\frac{$(a)}{$(b)\\sqrt{$(c)}}"),
                "Multiply by \\(\\frac{\\sqrt{$(c)}}{\\sqrt{$(c)}}\\): \\(\\frac{$(a)\\sqrt{$(c)}}{$(b) \\cdot $(c)}\\)",
                "Simplify: \\($(ans_str)\\)",
            ),
            time=120,
        )
        
    elseif problem_type == 3
        # Rationalizing with binomial denominators (1500-1700)
        # 1/(a + b√c) = (a - b√c)/(a² - b²c)
        a = randint(2, 8)
        b = randint(2, 6)
        c = choice([2, 3, 5, 7])
        
        denom_result = a^2 - b^2 * c
        
        # Make sure denominator doesn't become zero or too small
        while abs(denom_result) < 2
            a = randint(2, 8)
            b = randint(2, 6)
            denom_result = a^2 - b^2 * c
        end
        
        gcd_a = gcd(a, abs(denom_result))
        gcd_b = gcd(b, abs(denom_result))
        gcd_val = gcd(gcd_a, gcd_b)
        
        num_a = div(a, gcd_val)
        num_b = div(b, gcd_val)
        final_denom = div(denom_result, gcd_val)
        
        sign = final_denom < 0 ? "-" : ""
        final_denom = abs(final_denom)
        
        if final_denom == 1
            ans_str = "$(sign)$(num_a) - $(num_b)\\sqrt{$(c)}"
        else
            ans_str = "\\frac{$(sign)$(num_a) - $(num_b)\\sqrt{$(c)}}{$(final_denom)}"
        end
        
        problem(
            question="Rationalize the denominator: \\(\\frac{1}{$(a) + $(b)\\sqrt{$(c)}}\\)",
            answer=ans_str,
            difficulty=(1500, 1700),
            solution=steps(
                sol("Given", "\\frac{1}{$(a) + $(b)\\sqrt{$(c)}}"),
                "Multiply by conjugate: \\(\\frac{$(a) - $(b)\\sqrt{$(c)}}{($(a))^2 - ($(b))^2($(c))}\\)",
                "Simplify denominator: \\($(a^2) - $(b^2 * c) = $(denom_result)\\)",
                sol("Answer", ans_str),
            ),
            time=180,
        )
        
    elseif problem_type == 4
        # Solving radical equations (1400-1600)
        # √(ax + b) = c
        ans = randint(1, 25)
        a = randint(2, 8)
        b = randint(-15, 15)
        c = randint(3, 12)
        
        # Make sure c² = a*ans + b
        b = c^2 - a * ans
        
        problem(
            question="Solve for \\(x\\): \\(\\sqrt{$(a)x + $(b)} = $(c)\\)",
            answer=ans,
            difficulty=(1400, 1600),
            solution=steps(
                sol("Given", "\\sqrt{$(a)x + $(b)} = $(c)"),
                "Square both sides: \\($(a)x + $(b) = $(c^2)\\)",
                "Solve for \\(x\\): \\($(a)x = $(c^2 - b)\\)",
                sol("Answer", "x = $(ans)"),
            ),
            time=120,
        )
        
    elseif problem_type == 5
        # Adding/subtracting radicals (1200-1400)
        # a√c + b√c = (a+b)√c
        a = randint(2, 9)
        b = randint(2, 9)
        c = choice([2, 3, 5, 6, 7, 10, 11, 13, 15])
        sum_val = a + b
        
        problem(
            question="Simplify: \\($(a)\\sqrt{$(c)} + $(b)\\sqrt{$(c)}\\)",
            answer="$(sum_val)\\sqrt{$(c)}",
            difficulty=(1200, 1400),
            solution=steps(
                sol("Given", "$(a)\\sqrt{$(c)} + $(b)\\sqrt{$(c)}"),
                "Combine like terms (same radical)",
                sol("Answer", "$(sum_val)\\sqrt{$(c)}"),
            ),
            time=60,
        )
        
    else
        # Solving equations with two radicals (1600-1800)
        # √(ax + b) + √(cx + d) = e
        ans = randint(2, 20)
        a = randint(2, 6)
        c = randint(2, 6)
        val1 = randint(2, 8)
        val2 = randint(2, 8)
        e = val1 + val2
        
        b = val1^2 - a * ans
        d = val2^2 - c * ans
        
        problem(
            question="Solve for \\(x\\): \\(\\sqrt{$(a)x + $(b)} + \\sqrt{$(c)x + $(d)} = $(e)\\)",
            answer=ans,
            difficulty=(1600, 1800),
            solution=steps(
                sol("Given", "\\sqrt{$(a)x + $(b)} + \\sqrt{$(c)x + $(d)} = $(e)"),
                "Isolate one radical and square both sides",
                "After simplification and solving: \\(x = $(ans)\\)",
                "Verify by substituting back into original equation",
            ),
            time=240,
        )
    end
end